File: csharp/structural/composite.cs
namespace Gof.ModernCSharp.Structural;

// Composite treats individual objects and object groups through the same interface.
// Modern C# note: useful for trees such as menus, documents, ASTs, filesystems,
// and UI structures.
public static class Composite
{
    interface INode { int TotalSize(); }
    sealed record FileNode(string Name, int Size) : INode { public int TotalSize() => Size; }
    sealed class FolderNode(string name) : INode
    {
        readonly List<INode> _children = [];
        public void Add(INode child) => _children.Add(child);
        public int TotalSize() => _children.Sum(child => child.TotalSize());
        public override string ToString() => name;
    }

    public static string Run()
    {
        var root = new FolderNode("root");
        root.Add(new FileNode("readme.txt", 2));
        var assets = new FolderNode("assets");
        assets.Add(new FileNode("logo.png", 10));
        root.Add(assets);
        return root.TotalSize().ToString();
    }
}

File: py/structural/composite.py
# Composite treats individual objects and object groups through the same interface.
# Modern Python note: this remains useful for trees, documents, menus, filesystems,
# and UI structures where leaves and groups should share operations.
class File:
    def __init__(self, name, size):
        self.name = name
        self.size = size

    def total_size(self):
        return self.size


class Folder:
    def __init__(self, name):
        self.name = name
        self.children = []

    def add(self, child):
        self.children.append(child)

    def total_size(self):
        return sum(child.total_size() for child in self.children)


if __name__ == "__main__":
    root = Folder("root")
    root.add(File("readme.txt", 2))
    assets = Folder("assets")
    assets.add(File("logo.png", 10))
    root.add(assets)
    print(root.total_size())
